004-testing.md

```diff
@@ -0,0 +1,50 @@
+
+We [don't practice strict TDD](https://dhh.dk/2014/tdd-is-dead-long-live-testing.html) or pursue 100% test coverage. Instead, we try to get the most "bang for our buck" by writing [fewer higher-level integration tests](https://youtu.be/Fha2bVoC8SE) that test the real functionality that our application exposes, rather than isolated pieces.
+
+In [Phoenix](https://github.com/DoSomething/phoenix), this means the majority of our test coverage comes from [Cypress](https://www.cypress.io/) browser tests, which step through full application flows just like a real user. We also maintain some unit tests, written with [Jest](https://jestjs.io/) & [Enzyme](https://airbnb.io/enzyme/), for more complicated components. In backend Laravel applications, like [Northstar](https://github.com/DoSomething/northstar) and [Rogue](https://github.com/DoSomething/rogue), the majority of our test coverage comes from [PHPUnit](https://laravel.com/docs/5.5/http-tests) functional tests, which simulate real API requests.
+
+### Test Structure
+
+In general, all of our tests follow the same basic structure: "given this scenario, when I do X, I see Y as a result". For example,  ["given this JSON payload, the ContentfulEntry renders the appropriate component"](https://github.com/DoSomething/phoenix-next/blob/dc9713b568060a3a459c715ffb97854321d4ae10/resources/assets/components/ContentfulEntry/ContentfulEntry.test.js#L19-L26) or ["given a user isn't signed up for the campaign, if they click on the Join Now button, then they should see the affirmation modal"](https://github.com/DoSomething/phoenix-next/blob/dc9713b568060a3a459c715ffb97854321d4ae10/cypress/integration/campaign-signup.js#L35-L53).
+
+```js
+describe('Test Suite name', () => {
+   it('should do the things', () => {
+      // 1. Given some state, such as from a mock or factory...
+      // 2. When something happens (click, HTTP request, function call...)
+      // 3. Assert the result matches our expectations!
+   }
+});
+```
+
+### Process
+We like to write tests while adding features or fixing bugs, because writing tests help us "check our work" & demonstrate functionality in pull requests, and because otherwise it's easy to forget! We're still catching up on tests we skipped writing when we originally built Phoenix, Chompy, and other applications.
+
+We try not to merge in code without tests, and never merge in a pull request that introduces a test failure.
+
+### Mocking
+
+Since each of our applications are only small piece of our overall system, we use [mocks](https://en.wikipedia.org/wiki/Mock_object) to avoid relying on external dependencies in our tests. This helps to save us from "flakiness" (such as network issues) and allows us to easily simulate how applications would react to hypothetical scenarios (such as quickly testing how Phoenix's campaign landing page looks for a user with or without an existing campaign signup in Rogue).
```

Comment:
_Side note:_ Not 100% sure if some of our tests account for this or not, but maybe actually accounting for network issues in tests would be useful too.

For example, if Phoenix can't connect to Northstar for whatever reason, does it just stop in its tracks or is there some better user experience in place and maybe some tests that mimic this and check that the expected "fail" scenario is presented. 🤔

Reply:
Yeah, I don't think that we have any test coverage for this right now, but it's absolutely worth testing. Adding more tests like this could also go hand in hand with restarting fire drills (e.g. destroy Northstar QA offline one day & see how hard it is to recover), which is something we're hoping to do sometime this year!